cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -648,53 +649,69 @@ func (c *machineController) findScalableResourceProviderIDs(scalableResource *un
+		// nodes need to be filtered so that in a case where a machine is not becoming a node, or the instance
+		// lifecycle has changed during provisioning (eg spot instance going away), or the core autoscaler has
+		// decided that the node is not needed.
+		// Look for a node reference in the status, a machine without a node reference, and that is also not
+		// in failed or deleting state, has not yet become a node, and should be marked as pending.
+		// We give these machines normalized provider IDs to aid in the filtering process.
+		_, found, err = unstructured.NestedFieldCopy(machine.UnstructuredContent(), "status", "nodeRef")
+		if err != nil {
+			return nil, err
+		}
+
+		if !found {
+			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.GetName())
+			providerIDs = append(providerIDs, createPendingMachineProviderID(machine.GetNamespace(), machine.GetName()))
+			continue
+		}
+
+		// Running Machines
+		// We have filtered out the machines in failed, deleting, and pending states. We now check the provider
+		// ID and potentially the node reference details. It is ok for a machine not to have a provider ID as
+		// not all CAPI provider implement this field, but a machine in running state should have a valid
+		// node reference. If a provider ID is present, we add that to the list as we know it is not failed,
+		// deleting, or pending. If an empty provider ID is present, we check the node details to ensure that
+		// the machine references a valid node.
 		providerID, found, err := unstructured.NestedString(machine.UnstructuredContent(), "spec", "providerID")
 		if err != nil {
 			return nil, err
 		}
 
 		if found {
 			if providerID != "" {
-				// If the machine is deleting, prepend the deletion guard on the provider id
-				// so that it can be properly filtered when counting the number of nodes and instances.
-				if !machine.GetDeletionTimestamp().IsZero() {
-					klog.V(4).Infof("Machine %q has a non-zero deletion timestamp", machine.GetName())
-					providerID = createDeletingMachineNormalizedProviderID(machine.GetNamespace(), machine.GetName())
-				}
+				// Machine has a provider ID, add it to the list
 				providerIDs = append(providerIDs, providerID)
 				continue
 			}
 		}
 
 		klog.Warningf("Machine %q has no providerID", machine.GetName())
 
-		// Look for a node reference in the status, a machine with a provider ID but no node reference has not
-		// yet become a node and should be marked as pending.
-		_, found, err = unstructured.NestedFieldCopy(machine.UnstructuredContent(), "status", "nodeRef")
-		if err != nil {
-			return nil, err
-		}
-
-		if !found {
-			klog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.GetName())
-			providerIDs = append(providerIDs, createPendingMachineProviderID(machine.GetNamespace(), machine.GetName()))
-			continue
-		}
-
+		// Begin checking to determine if the node reference is valid
```

Comment:
does the `found` response from `unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "kind")` (L715) need to be taken into consideration before we call `unstructured.NestedString(machine.UnstructuredContent(), "status", "nodeRef", "name")` (L725)?

I see we're considering it in L720 (`if found && nodeRefKind != "Node" {`), but is that the only place we want to evaluate it?

Reply:
yes, i think we just need to check it that one time. i left this part of the logic unchanged as it seemed to be correct.